File: src/Niche/Prolog/decorator.pl
base(alert).

audit(Component, Decorated) :-
    format(atom(Decorated), 'audit(~w)', [Component]).

encrypt(Component, Decorated) :-
    format(atom(Decorated), 'enc(~w)', [Component]).

run :-
    base(Base),
    audit(Base, Audited),
    encrypt(Base, Encrypted),
    encrypt(Base, Inner),
    audit(Inner, Stacked),
    format('base=~w~n', [Base]),
    format('audit=~w~n', [Audited]),
    format('encrypted=~w~n', [Encrypted]),
    format('stacked=~w~n', [Stacked]).

File: src/DataScience/MATLAB/decorator.m
function decorator()
    base = @() "alert";
    audit = @(inner) @() "audit(" + inner() + ")";
    encrypt = @(inner) @() "enc(" + inner() + ")";

    audited = audit(base);
    encrypted = encrypt(base);
    stacked = audit(encrypt(base));

    fprintf('base=%s\n', base());
    fprintf('audit=%s\n', audited());
    fprintf('encrypted=%s\n', encrypted());
    fprintf('stacked=%s\n', stacked());
end
